File: python/runkite_runner/run_status.py
"""Pre-execution status check (Runner Protocol §10.3).

After GetJob dequeues a RunAssignment, the runner MUST ask the control
plane whether the run is still executable before starting agent work --
a cancel that landed between dequeue and execute is otherwise a silent
race. Fail-open on transport errors: WatchCancels remains the primary
cancel path; a brief CP blip must not drop every job.
"""

from __future__ import annotations

import logging

import httpx

from .tls_utils import httpx_tls_kwargs

logger = logging.getLogger("runkite.runner")

# Terminal + interrupted: MUST NOT execute (PROTOCOL §10.3).
_SKIP_STATUSES = frozenset({"interrupted", "success", "error", "timeout"})


async def should_skip_run(
    http_address: str,
    run_id: str,
    *,
    runner_kind: str = "",
    runner_token: str = "",
) -> bool:
    """Return True when the assignment must be discarded without executing."""
    base = http_address.rstrip("/")
    headers: dict[str, str] = {}
    if runner_token:
        headers["X-Runner-Kind"] = runner_kind or "python-langgraph"
        headers["X-Runner-Token"] = runner_token
    try:
        async with httpx.AsyncClient(timeout=5.0, **httpx_tls_kwargs()) as client:
            resp = await client.get(f"{base}/internal/runs/{run_id}/status", headers=headers)
        if resp.status_code == 404:
            logger.warning("pre-exec status: run %s not found; discarding", run_id)
            return True
        if resp.status_code != 200:
            logger.warning(
                "pre-exec status: run %s HTTP %s; proceeding (fail-open)",
                run_id,
                resp.status_code,
            )
            return False
        status = (resp.json() or {}).get("status", "")
        if status in _SKIP_STATUSES:
            logger.info("pre-exec status: run %s is %s; discarding assignment", run_id, status)
            return True
        return False
    except Exception as e:
        logger.warning("pre-exec status: run %s check failed (%s); proceeding (fail-open)", run_id, e)
        return False

File: typescript/runkite-runner/src/runStatus.ts
/**
 * Pre-execution status check (Runner Protocol §10.3).
 * After GetJob dequeues a RunAssignment, ask the control plane whether
 * the run is still executable before starting agent work. Fail-open on
 * transport errors -- WatchCancels remains the primary cancel path.
 */
import { logger } from "./logger.js";
import { httpDispatcher } from "./tls.js";

const SKIP_STATUSES = new Set(["interrupted", "success", "error", "timeout"]);

type FetchInit = RequestInit & { dispatcher?: unknown };

export async function shouldSkipRun(
  httpAddress: string,
  runId: string,
  opts?: { runnerKind?: string; runnerToken?: string },
): Promise<boolean> {
  const base = httpAddress.replace(/\/+$/, "");
  const headers: Record<string, string> = {};
  if (opts?.runnerToken) {
    headers["X-Runner-Kind"] = opts.runnerKind || "typescript-langgraphjs";
    headers["X-Runner-Token"] = opts.runnerToken;
  }
  try {
    const init: FetchInit = { headers, dispatcher: httpDispatcher() };
    const resp = await fetch(`${base}/internal/runs/${runId}/status`, init);
    if (resp.status === 404) {
      logger.warn(`pre-exec status: run ${runId} not found; discarding`);
      return true;
    }
    if (!resp.ok) {
      logger.warn(`pre-exec status: run ${runId} HTTP ${resp.status}; proceeding (fail-open)`);
      return false;
    }
    const body = (await resp.json()) as { status?: string };
    const status = body.status ?? "";
    if (SKIP_STATUSES.has(status)) {
      logger.info(`pre-exec status: run ${runId} is ${status}; discarding assignment`);
      return true;
    }
    return false;
  } catch (err) {
    logger.warn(`pre-exec status: run ${runId} check failed (${err}); proceeding (fail-open)`);
    return false;
  }
}
